compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,238 @@
+      Value toSIMD = rewriter.create<IREE::VectorExt::ToSIMDOp>(
+          oldResult.getLoc(), oldResult.getType(), replacement);
+      // Add to replacements.
+      replacement = toSIMD;
+    }
+    replacements.push_back(replacement);
+  }
+
+  rewriter.replaceOp(op, replacements);
+}
+
+std::optional<DistributionSignature>
+DistributionPattern::getOpSignature(Operation *op) const {
+  if (!hasOpSignature(op)) {
+    return std::nullopt;
+  }
+  return ::mlir::iree_compiler::getOpSignature(op);
+}
+
+LogicalResult DistributionPattern::match(Operation *op) const {
+  if (!hasOpSignature(op)) {
+    return failure();
+  }
+  return match(op);
+}
+
+static void
+debugPrintUniqueOperationNames(SmallVectorImpl<Operation *> &worklist) {
+  DenseSet<StringRef> uniqueNames;
+  for (Operation *op : worklist) {
+    uniqueNames.insert(op->getName().getStringRef());
+  }
+
+  for (StringRef name : uniqueNames) {
+    llvm::dbgs().indent(2) << "* " << name << "\n";
+  }
+  LLVM_DEBUG(llvm::dbgs() << "\n");
+}
+
+/// A rewriter for the pattern rewriting driver.
+class VectorDistributionRewriter : public PatternRewriter {
+public:
+  VectorDistributionRewriter(MLIRContext *ctx) : PatternRewriter(context) {}
+};
+
+static void applyVectorDistribution(Operation *root,
+                                    const FrozenRewritePatternSet &patterns) {
+
+  SmallVector<Operation *> worklist;
+
+  VectorDistributionRewriter rewriter(root->getContext());
+  PatternApplicator applicator(patterns);
+  applicator.applyDefaultCostModel();
+
+  // Collect all the operations to be distributed.
+  LLVM_DEBUG(llvm::dbgs() << "Collecting operations to be distributed\n");
+  root->walk([&](Operation *op) {
+    if (hasOpSignature(op)) {
+      worklist.push_back(op);
+    }
+  });
+  LLVM_DEBUG(llvm::dbgs() << "Operations to be distributed:\n");
+  LLVM_DEBUG(debugPrintUniqueOperationNames(worklist));
+
+  // Note that the pattern application here never runs on a newly created
+  // operation. It always runs on an existing operation. This ensures that no
+  // invalidated state of the analysis is ever used.
+  for (Operation *op : worklist) {
+    LLVM_DEBUG(llvm::dbgs() << "Distributing: ");
+    LLVM_DEBUG(op->print(llvm::dbgs(), OpPrintingFlags().skipRegions()));
+    LLVM_DEBUG(llvm::dbgs() << "\n");
+    if (failed(applicator.matchAndRewrite(op, rewriter))) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << ": Failed to distribute operation:\n");
+      continue;
+    }
+
+    LLVM_DEBUG(llvm::dbgs().indent(2)
+               << ": Successfully distributed operation:\n");
+  }
+}
+
+static bool canDistribute(Operation *op, VectorLayoutAnalysis &analysis) {
+  auto values = llvm::to_vector_of<Value>(op->getOperands());
+  llvm::append_range(values, op->getResults());
+
+  // Check if all operands and results of this operation have a layout.
+  return llvm::all_of(values, [&](Value value) -> bool {
+    if (auto vectorValue = dyn_cast<VectorValue>(value)) {
```

Comment:
Nit: I think this can just be

```
auto vectorValue = dyn_cast<VectorValue>(value);
return vectorValue && analysis.getLayout<Attribute>(vectorValue);
```